flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpillingAdaptiveSpanningRecordDeserializer.java

```diff
@@ -95,13 +96,14 @@ public Buffer getCurrentBuffer () {
 	}
 
 	@Override
-	public Optional<Buffer> getUnconsumedBuffer() {
+	public Optional<Buffer> getUnconsumedBuffer() throws IOException {
+		Optional<MemorySegment> target;
 		if (nonSpanningWrapper.remaining() > 0) {
-			return Optional.of(new NetworkBuffer(nonSpanningWrapper.copyToTargetSegment(), FreeingBufferRecycler.INSTANCE));
+			target = nonSpanningWrapper.getUnconsumedSegment();
 		} else {
-			MemorySegment target = spanningWrapper.copyToTargetSegment();
-			return target != null ? Optional.of(new NetworkBuffer(target, FreeingBufferRecycler.INSTANCE)) : Optional.empty();
+			target = spanningWrapper.getUnconsumedSegment();
 		}
+		return target.map(memorySegment -> new NetworkBuffer(memorySegment, FreeingBufferRecycler.INSTANCE, true, memorySegment.size()));
```

Comment:
Actually thought about doing the same thing ;)

Reply:
:)